azurelinuxagent/common/utils/processutil.py

```diff
@@ -123,36 +120,34 @@ def capture_from_process_poll(process, cmd, timeout, code):
     if return_code != 0:
         raise ExtensionError("Non-zero exit code: {0}, {1}".format(return_code, cmd), code=code)
 
-    stderr = b''
-    stdout = b'cannot collect stdout'
-
-    # attempt non-blocking process communication to capture output
-    def proc_comm(_process, _return):
-        try:
-            _stdout, _stderr = _process.communicate()
-            _return[0] = _stdout
-            _return[1] = _stderr
-        except Exception:
-            pass
-
-    try:
-        mgr = multiprocessing.Manager()
-        ret_dict = mgr.dict()
-
-        cproc = Process(target=proc_comm, args=(process, ret_dict))
-        cproc.start()
-
-        # allow 1s to capture output
-        cproc.join(1)
-
-        if len(ret_dict) == 2:
-            stdout = ret_dict[0]
-            stderr = ret_dict[1]
-
-    except Exception:
-        pass
-
-    return stdout, stderr
+    # at this point the parent process has completed and we need to pickup its output; we use select()
+    # with a short timeout to capture the beginning of the output of any child processes.
```

Comment:
If you wait until child process exit to consume stdout or stderr, you WILL deadlock if the child produces more output than will fit in the socket buffer. There _is_ a limit. When the child process does a write to a full socket, the write() syscall will block; the child process will never exit; the agent will never see process death and will never drain the socket.

There were good reasons for the previous design. This is a bad change.

Reply:
@jasonzio When you say _child process exit_, do you mean the child process of the `ext_handler` (which is the extension code) or the child of the extension itself (fork from an extension, a la custom script approach).
- If it is the extension itself, the code at line#119 is already a blocking call.
- And for the forked process of the extension, the earlier approach was not taking care of cases when the daemon was detaching the parent process (an extension which starts a daemon, but did not successfully detach itself). Thus when it did the `_process.communicate()` in the `proc_comm(_process, _return)`, it would not exit as well ([communicate](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.communicate) being a blocking call).

Reply:
I still need to verify the extension itself doesn't block... if this turns out to be problematic the other option we are considering is stop capturing output altogether. The current approach of creating a subprocess that calls communicate() is also producing hangs in the agent. I will update this thread.